### ISSUE TITLE
Using Rpmalloc for a Garbage Collector

### ISSUE THREAD

**Issue by ncannasse** [mjansson/rpmalloc#140]
As part for HashLink VM (https://github.com/haxeFoundation/hashlink) we are looking at a better allocator for our garbage collector.

We have started some refactor effort to separate the core GC from the allocator, and we are left with the following allocator API:

```C
// Initialize the allocator
void gc_allocator_init();

// Get the block size within the given page. The block validity has already been checked.
int gc_allocator_fast_block_size( gc_pheader *page, void *block );

// Get the block id within the given page, or -1 if it's an invalid ptr.
// The block is already checked within page bounds
int gc_allocator_get_block_id( gc_pheader *page, void *block );

// Allocate a block with given size using the specified page kind. 
// Returns NULL if no block could be allocated
// Sets size to really allocated size (could be larger)
// Sets size to -1 if allocation refused (required size is invalid)
void *gc_allocator_alloc( int *size, int page_kind );
```

And the allocator wrapper can callback the following in our GC to perform process pages allocation and free, and is allowed to store custom data on a per "page" basis (one page is a multiple of 64KB contiguous memory space):

```C
gc_pheader *gc_alloc_page( int size, int kind, int block_count );
void gc_free_page( gc_pheader *page, int block_count );
```

Few questions regarding rpmalloc:
- is it possible from a given pointer to know if the pointer is valid, and get its size ? we don't allow internal pointers so we should be sure that the pointer is currently valid. Ofc this should be done very efficiently, in constant time.
- I see that rpmalloc is using some fixed size classes, but is there a way to control the mapping between classes and pages in order to ensure that the same page does not contain different classes, and maintain up-to-date page metadata with class information, this would allow to calculate the block "id", which is used to perform marking (one bit per block, X bytes per page)

**Comment by mjansson** [comment]
rpmalloc does not provide a way to check pointer validity, as this would require keeping track of memory pages to be able to look up if a pointer falls inside a tracked memory page.

It does, however, provide a very fast constant time size lookup, as it basically comes down to masking pointer address to reach the memory page span header and lookup the block size class.

In the context of rpmalloc all blocks within a "span" of memory pages contain blocks of the same size. A span can never contain blocks of different sizes. However, since rpmalloc supports huge pages it is possible that a single memory page (for example if using 1GiB memory pages) can contain multiple "spans", and thus blocks of different sizes.

**Comment by ncannasse** [comment]
> rpmalloc does not provide a way to check pointer validity, as this would require keeping track of memory pages to be able to look up if a pointer falls inside a tracked memory page.

Our GC already maintains a mapping of allocated pages, so it knows if a pointer is accessible. However it needs to tell if a pointer is not a false positive (when scanning stack in particular). So in our current (slow) allocator we bind metadata to each "span" (which is a set of pages) to tell which block size are used. 

> It does, however, provide a very fast constant time size lookup, as it basically comes down to masking pointer address to reach the memory page span header and lookup the block size class.

If we disable huge pages support, does it means that we can read the span header for each 64KB page to know the size of the blocks it contains ? What about large blocks ?

**Comment by mjansson** [comment]
Yes, the span header is always at the start of the span of memory pages, and by default spans are 64KiB in size.

The same applies to large blocks, with the limit between small/medium blocks and large blocks being half a span size, or rather ((span_size - span_header_size) / 2) in more detail. However, since a large block can be more than one span in size, you can only query the size using the pointer to the start of the block, not any memory address within the block.

It also applies to huge blocks (more than 64 spans), but in this case the header counts the size in number of memory pages rather than spans, since the implementation directly maps/unmaps these huge blocks and does no attempt at caching within the allocator.

So to summarize, as long as you have a pointer to the start of a memory block, you can always mask the address with the span alignment (default 64KiB) to get the span header, then read out the meta data to find block size there. Which is exactly what _memory_usable_size does - https://github.com/mjansson/rpmalloc/blob/797132c5130ab35a52f99b4948f3bbccefd24730/rpmalloc/rpmalloc.c#L1735
